crates/primitives/oracle/src/pragma.rs

```diff
@@ -0,0 +1,135 @@
+use std::fmt;
+
+use anyhow::{bail, Context};
+use serde::{Deserialize, Serialize};
+
+pub const DEFAULT_API_URL: &str = "https://api.dev.pragma.build/node/v1/data/";
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PragmaOracle {
+    #[serde(default = "default_oracle_api_url")]
+    pub api_url: String,
+    #[serde(default)]
+    pub api_key: String,
+    #[serde(default)]
+    pub aggregation_method: AggregationMethod,
+    #[serde(default)]
+    pub interval: Interval,
+    #[serde(default)]
+    pub price_bounds: PriceBounds,
+}
+
+impl Default for PragmaOracle {
+    fn default() -> Self {
+        Self {
+            api_url: default_oracle_api_url(),
+            api_key: String::default(),
+            aggregation_method: AggregationMethod::Median,
+            interval: Interval::OneMinute,
+            price_bounds: Default::default(),
+        }
+    }
+}
+
+impl PragmaOracle {
+    pub fn new(api_url: String, api_key: String) -> Self {
+        Self {
+            api_url,
+            api_key,
+            aggregation_method: AggregationMethod::Median,
+            interval: Interval::OneMinute,
+            price_bounds: Default::default(),
+        }
+    }
+
+    pub fn get_fetch_url(&self, base: String, quote: String) -> String {
+        format!("{}{}/{}?interval={}&aggregation={}", self.api_url, base, quote, self.interval, self.aggregation_method)
+    }
+
+    pub async fn fetch_eth_strk_price(&self) -> anyhow::Result<(u128, u32)> {
+        let response = reqwest::Client::new()
+            .get(self.get_fetch_url(String::from("eth"), String::from("strk")))
+            .header("x-api-key", self.api_key.clone())
+            .send()
+            .await
+            .context("failed to retrieve price from pragma oracle")?;
+
+        let oracle_api_response = response.json::<PragmaApiResponse>().await.context("failed to parse api response")?;
+        let eth_strk_price = u128::from_str_radix(oracle_api_response.price.trim_start_matches("0x"), 16).context("failed to parse price")?;
+        if eth_strk_price ==  0 {
+            bail!("Pragma api returned 0 for eth/strk price");
+        }
+        Ok((eth_strk_price, oracle_api_response.decimals))
+    }
+}
+
+#[derive(Default, Debug, Serialize, Deserialize, Clone)]
+/// Supported Aggregation Methods
+pub enum AggregationMethod {
+    #[serde(rename = "median")]
+    Median,
+    #[serde(rename = "mean")]
+    Mean,
+    #[serde(rename = "twap")]
+    #[default]
+    Twap,
+}
```

Comment:
Maybe some short comments on what each one of these do? `Median` and `Mean` seem pretty self-explanatory but `Twap` a little less so.

Reply:
I can if its mandatory but I don't know if its relevant as twap is also a pretty common aggregation method in the financial world.

Reply:
`#[serde(rename_all = "snake_case")]` You can simplify the renaming by using this attribute at the enum level, instead of applying `#[serde(rename)]` to each variant individually. This ensures consistency and reduces boilerplate.

Reply:
imo, `Twap` is self-explanatory. Just add a doc comment explaining the acronym